src/constant.jl

```diff
@@ -5,54 +5,47 @@
 export Constant
 export vexity, evaluate, sign, conic_form!
 
-struct Constant <: AbstractExpr
+const ComplexValue = Union{Complex,AbstractVecOrMat{<:Complex}}
+
+_size(x::Number) = (1, 1)
+_size(x::AbstractVector) = (length(x), 1)
+_size(x::AbstractMatrix) = size(x)
+
+_sign(x::ComplexValue) = ComplexSign()
+_sign(x::Number) = x >= 0 ? Positive() : Negative()
+_sign(v::AbstractVecOrMat) = all(x->x >= 0, v) ? Positive() : Negative()
+
+struct Constant{T<:Value} <: AbstractExpr
     head::Symbol
     id_hash::UInt64
-    value::Value
+    value::T
     size::Tuple{Int, Int}
-    vexity::Vexity
     sign::Sign
 
-    function Constant(x::Value, sign::Sign)
-        sz = (size(x, 1), size(x, 2))
-        return new(:constant, objectid(x), x, sz, ConstVexity(), sign)
+    function Constant(x::T, sign::Union{ComplexSign,NoSign}) where T<:ComplexValue
+        new{T}(:constant, objectid(x), x, _size(x), sign)
     end
 
-    function Constant(x::Value, check_sign::Bool=true)
-        if check_sign
-            if !isreal(x)
-                return Constant(x, ComplexSign())
-            elseif all(xi >= 0 for xi in x)
-                return Constant(x, Positive())
-            elseif all(xi <= 0 for xi in x)
-                return Constant(x, Negative())
-            end
-        end
-        return Constant(x, NoSign())
+    function Constant(x::T, sign::Union{Positive,Negative,NoSign}) where T<:Value
```

Comment:
While likely uncommon, I just realized that we could(/should?) support something like
```julia
Constant(x::Real, sign::ComplexSign) = Constant(Complex(x), sign)
```

Reply:
So currently Convex is fine with whatever sign you tell it that a `Constant` has and it stores the value with no conversion or checking against the given sign. This PR removes that entirely; mismatched value types and signs result in `MethodError`s. I'm not sure whether that's a bug or feature.